### ISSUE TITLE
Destructuring assignment doesn't work for bare unit structs

### ISSUE THREAD

**Issue by jsgf** [rust-lang/rust#118753]
I tried this code:

```rust
struct Unit;

Unit = Unit;
```
I expected to see this being treated as a (trivial) destructuring assignment.

Instead, it fails with:
```
error[E0070]: invalid left-hand side of assignment
 --> src/main.rs:3:10
  |
3 |     Unit = Unit;
  |     ---- ^
  |     |
  |     cannot assign to this expression

For more information about this error, try `rustc --explain E0070`.
error: could not compile `playground` (bin "playground") due to previous error
```

Likewise:
```rust
enum Enum { Unit }

Enum::Unit = Enum::Unit;
```
fails similarly.

On the other hand the very similar:
```rust
struct Tup0();

Tup0() = Tup0();
```
works as expected.

The [RFC](https://rust-lang.github.io/rfcs/2909-destructuring-assignment.html) for this says:
> We support the following classes of expressions:
> - Tuples.
> - Slices.
> - Structs (including unit and tuple structs).
> - Unique variants of enums.

It looks like enums are also not implemented. The [PR](https://github.com/rust-lang/rust/pull/78836) implementing this for structs seems to overlook unit structs, and I couldn't find any place claiming to implement it for enums.

### Meta

`rustc --version --verbose`:
```
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: x86_64-unknown-linux-gnu
release: 1.71.1
LLVM version: 16.0.5
```

And is also present in nightly.

**Comment by jsgf** [comment]
Oh, this was supposed to have been fixed in [1.62](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1620-2022-06-30) with https://github.com/rust-lang/rust/pull/95380/

**Comment by jsgf** [comment]
```
Unit {} = Unit;
```
*does* work, using the "`{}` can be used on any unit struct" loophole.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ee4970582dd71931f907fc9b092c7d73 explores the matrix of different unit types in different contexts and highlights a number of inconsistencies.
